app/controllers/amr_heatmap_controller.rb

```diff
@@ -58,4 +63,174 @@ def amr_counts
+      get_s3_file(S3_CARD_OWL)
+    end
+    return card_owl
+  end
+
+  def search_card_owl(gene_name, xml_dom)
+    search_result = {}
+    err = "No match found for %s in the CARD Antibiotic Resistance Ontology."
+    owl_classes = xml_dom.xpath(".//owl:Class")
+    matching_entry = nil
+
+    # normalize gene name
+    alphanumeric_gene_name = gene_name.downcase.gsub(/\W/, '')
+    regex_for_gene = Regexp.new("\\b#{alphanumeric_gene_name}\\b")
+    # search through owl classes
+    owl_classes.each do |owl_class|
+      # first we check the label
+      label = owl_class.at_xpath("./rdfs:label").content
+      separated_label = label.downcase.tr('/', ' ')
+      alphanumeric_label = separated_label.gsub(/[^0-9a-z_\s]/, '')
+      if regex_for_gene.match?(alphanumeric_label)
+        matching_entry = owl_class
+        break
+      end
+
+      # then we check the synonyms
+      synonyms = owl_class.xpath("./oboInOwl:hasExactSynonym")
+      synonyms.each do |synonym|
+        separated_synonym = synonym.content.downcase.tr('/', ' ')
+        alphanumeric_synonym = separated_synonym.gsub(/[^0-9a-z_\s]/, '')
+        if regex_for_gene.match?(alphanumeric_synonym)
+          matching_entry = owl_class
+          break
+        end
+      end
+    end
+
+    # abort if no matching entry
+    if matching_entry.nil?
+      search_result["error"] = format(err, gene_name)
+    else
+      search_result["matching_node"] = matching_entry
+    end
+
+    return search_result
+  end
+
+  def gather_ontology_information(matching_entry, xml_dom)
+    ontology_information = {}
+    # Gathering properties
+    ontology_information["label"] = matching_entry.at_xpath("./rdfs:label").content
+    ontology_information["accession"] = matching_entry.at_xpath("./oboInOwl:id").content.split(':')[1]
+    ontology_information["description"] = matching_entry.at_xpath("./obo:IAO_0000115").content
+
+    # These properties are referred to by accession number in the matching OWL Class,
+    # but their labels are located elsewhere in the document.
+    remote_properties = {
+      "drugClass" => get_drug_classes(ontology_information["accession"], xml_dom),
+      "geneFamily" => matching_entry.xpath("./rdfs:subClassOf[not(*)]")
+    }
+
+    # Some drug resistances are only listed in the gene family entry
+    if remote_properties["drugClass"].empty?
+      drug_classes = []
+      remote_properties["geneFamily"].each do |node|
+        gene_family_accession = node.get_attribute("rdf:resource").split('_')[1]
+        drug_classes.concat(get_drug_classes(gene_family_accession, xml_dom))
+      end
+      remote_properties["drugClass"] = drug_classes
+    end
+
+    local_properties = {
+      "synonyms" => matching_entry.xpath("./oboInOwl:hasExactSynonym"),
+      "publications" => matching_entry.at_xpath("./following-sibling::owl:Axiom").xpath("./oboInOwl:hasDbXref")
+    }
+
+    local_properties.each do |ontology_key, property_nodes|
+      collection = []
```

Comment:
you can actually just map into `ontology_information[ontology_key]` instead of needing the intermediate `collection = []`

Reply:
So `property_nodes` is actually a `NodeSet`, from Nokogiri, so map isn't available unfortunately :(